syz-verifier/main.go

```diff
@@ -0,0 +1,310 @@
+	for _, id := range cfg.Syscalls {
+		calls[target.Syscalls[id]] = true
+	}
+
+	vrf := &Verifier{
+		workdir:     workdir,
+		crashdir:    crashdir,
+		pools:       pools,
+		target:      target,
+		choiceTable: target.BuildChoiceTable(nil, calls),
+		rnd:         rand.New(rand.NewSource(time.Now().UnixNano() + 1e12)),
+		runnerBin:   runnerBin,
+		executorBin: execBin,
+		addr:        addr,
+	}
+
+	srv, err := startRPCServer(vrf)
+	if err != nil {
+		log.Fatalf("failed to initialise RPC server: %v", err)
+	}
+
+	for idx, pi := range pools {
+		go func(pi *poolInfo, idx int) {
+			for {
+				// TODO: implement support for multiple VMs per Pool.
+				inst, err := pi.pool.Create(0)
+				if err != nil {
+					log.Fatalf("failed to create instance: %v", err)
+				}
+				defer inst.Close()
+
+				fwdAddr, err := inst.Forward(srv.port)
+				if err != nil {
+					log.Fatalf("failed to set up port forwarding: %v", err)
+
+				}
+
+				runnerBin, err := inst.Copy(vrf.runnerBin)
+				if err != nil {
+					log.Fatalf(" failed to copy runner binary: %v", err)
+				}
+				_, err = inst.Copy(vrf.executorBin)
+				if err != nil {
+					log.Fatalf("failed to copy executor binary: %v", err)
+				}
+
+				cmd := instance.RunnerCmd(runnerBin, fwdAddr, vrf.target.OS, vrf.target.Arch, idx, 0)
+				outc, errc, err := inst.Run(pi.cfg.Timeouts.VMRunningTime, vrf.vmStop, cmd)
+				if err != nil {
+					log.Fatalf("failed to start runner: %v", err)
+				}
+
+				inst.MonitorExecution(outc, errc, pi.Reporter, vm.ExitTimeout)
+				srv.cleanup(idx, 0)
+			}
+		}(pi, idx)
+	}
+
+	select {}
+}
+
+func startRPCServer(vrf *Verifier) (*RPCServer, error) {
+	srv := &RPCServer{
+		vrf:   vrf,
+		pools: vrf.pools,
+		progs: make(map[int]*programInfo),
+	}
+
+	s, err := rpctype.NewRPCServer(vrf.addr, "Verifier", srv)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("serving rpc on tcp://%v", s.Addr())
+	srv.port = s.Addr().(*net.TCPAddr).Port
+
+	go s.Serve()
+	return srv, nil
+}
+
+// Connect notifies the RPCServer that a new Runner was started.
+func (srv *RPCServer) Connect(a *rpctype.RunnerConnectArgs, r *int) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	pool, vm := a.Pool, a.VM
+	srv.pools[pool].vmRunners[vm] = []int{}
+	return nil
+}
+
+// NewExchange is called when a Runner requests a new program to execute and, potentially, wants to send a new Result to the RPCServer.
+func (srv *RPCServer) NewExchange(a *rpctype.NewExchangeArgs, r *rpctype.NewExchangeRes) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	if a.Info.Calls != nil {
+		res := &verf.Result{
+			Pool:    a.Pool,
+			ProgIdx: a.ProgIdx,
+			Hanged:  a.Hanged,
+			Info:    a.Info,
+		}
+		if resReady := srv.newResult(res); resReady {
+			srv.sendForVerification(a.ProgIdx)
+		}
+	}
+
+	prog, pi := srv.newProgram(a.Pool, a.VM)
+	r.RPCProg = rpctype.RPCProg{Prog: prog, ProgIdx: pi}
+	return nil
+}
+
+func (srv *RPCServer) newProgram(pool int, vm int) ([]byte, int) {
```

Comment:
The same comment for use of int indices here.
You use `pool` here like 7 times and _always_ to convert it to the pool object `srv.pools[pool]`. If you turn the int into pool object pointer and early as possible and pass it around, it will make simpler and nicer.